File: taurus-protect-sdk-typescript/src/services/group-service.ts
/**
 * Group service for Taurus-PROTECT SDK.
 *
 * Provides methods for group management operations.
 */

import { NotFoundError, ValidationError } from '../errors';
import type { GroupsApi } from '../internal/openapi/apis/GroupsApi';
import { groupFromDto, groupsFromDto } from '../mappers/user';
import {
  buildOffsetPagination,
  offsetRequest,
  type PaginatedResult,
} from '../models/pagination';
import type { Group, ListGroupsOptions } from '../models/user';
import { BaseService } from './base';
import { offsetQuery } from './paging';

/**
 * GetGroups computes enforcedInRules for each group and its users, and validatord omits a
 * false bool, so absent means false.
 */
function withComputedRulesFlags(group: Group): Group {
  return {
    ...group,
    enforcedInRules: group.enforcedInRules ?? false,
    users: group.users?.map((user) => ({
      ...user,
      enforcedInRules: user.enforcedInRules ?? false,
    })),
  };
}

/**
 * Service for group management operations.
 *
 * Provides methods to list and retrieve groups.
 *
 * @example
 * ```typescript
 * // List groups
 * const result = await groupService.list({ limit: 50 });
 * for (const group of result.items) {
 *   console.log(`${group.name}: ${group.userIds?.length ?? 0} users`);
 * }
 *
 * // Get single group
 * const group = await groupService.get('group-123');
 * console.log(`Group: ${group.name}`);
 * ```
 */
export class GroupService extends BaseService {
  private readonly groupsApi: GroupsApi;

  /**
   * Creates a new GroupService instance.
   *
   * @param groupsApi - The GroupsApi instance from the OpenAPI client
   */
  constructor(groupsApi: GroupsApi) {
    super();
    this.groupsApi = groupsApi;
  }

  /**
   * Gets a group by ID.
   *
   * @param groupId - The group ID to retrieve
   * @returns The group
   * @throws {@link ValidationError} If groupId is invalid
   * @throws {@link NotFoundError} If group not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const group = await groupService.get('group-123');
   * console.log(`Group: ${group.name}, Members: ${group.userIds?.length}`);
   * ```
   */
  async get(groupId: string): Promise<Group> {
    if (!groupId || groupId.trim() === '') {
      throw new ValidationError('groupId is required');
    }

    return this.execute(async () => {
      // The API doesn't have a direct get-by-id endpoint,
      // so we use the list endpoint with id filter
      const response = await this.groupsApi.userServiceGetGroups({
        limit: '1',
        offset: '0',
        ids: [groupId],
      });

      const result = response.result;

      if (!result || result.length === 0) {
        throw new NotFoundError(`Group ${groupId} not found`);
      }

      const group = groupFromDto(result[0]);
      if (!group) {
        throw new NotFoundError(`Group ${groupId} not found`);
      }

      return withComputedRulesFlags(group);
    });
  }

  /**
   * Lists groups with pagination.
   *
   * @param options - Optional filtering and pagination options
   * @returns Paginated result containing groups and pagination info
   * @throws {@link ValidationError} If limit or offset are invalid
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // List first 50 groups
   * const result = await groupService.list({ limit: 50 });
   * console.log(`Found ${result.pagination.totalItems} groups`);
   *
   * // Search groups
   * const result = await groupService.list({ query: 'admin' });
   * ```
   */
  async list(options?: ListGroupsOptions): Promise<PaginatedResult<Group>> {
    const page = offsetRequest(options);

    return this.execute(async () => {
      const response = await this.groupsApi.userServiceGetGroups({
        ...offsetQuery(page),
        ids: options?.ids,
        query: options?.query,
      });

      const rows = response.result ?? [];
      return {
        items: groupsFromDto(rows).map(withComputedRulesFlags),
        // The server may append a synthetic technical group beyond `limit`.
        pagination: buildOffsetPagination('plus_min_rows_limit', page, response, rows.length),
      };
    });
  }
}

File: taurus-protect-sdk-python/taurus_protect/services/group_service.py
"""Group service for Taurus-PROTECT SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, List, Optional, Tuple

from taurus_protect.mappers.user import group_from_dto, groups_from_dto
from taurus_protect.models.pagination import (
    PLUS_MIN_ROWS_LIMIT,
    Pagination,
    offset_pagination,
    offset_query,
    resolve_offset,
    resolve_page_size,
)
from taurus_protect.models.user import Group
from taurus_protect.services._base import BaseService

if TYPE_CHECKING:
    pass  # For OpenAPI types when available


def _with_computed_flags(group: Group) -> Group:
    """validatord omits a false bool, so where it computed the flags an absent one is False."""
    return group.model_copy(
        update={
            "enforced_in_rules": bool(group.enforced_in_rules),
            "users": [
                u.model_copy(update={"enforced_in_rules": bool(u.enforced_in_rules)})
                for u in group.users
            ],
        }
    )


class GroupService(BaseService):
    """
    Service for group management operations.

    Provides methods to list and retrieve groups.

    Example:
        >>> # List groups
        >>> groups, pagination = client.groups.list(limit=100, offset=0)
        >>> for group in groups:
        ...     print(f"{group.name}: {len(group.users)} users")
        >>>
        >>> # Get single group
        >>> group = client.groups.get("group-123")
        >>> print(f"Group: {group.name}")
    """

    def __init__(self, api_client: Any, groups_api: Any) -> None:
        """
        Initialize group service.

        Args:
            api_client: The OpenAPI client instance.
            groups_api: The GroupsApi service from OpenAPI client.
        """
        super().__init__(api_client)
        self._groups_api = groups_api

    def get(self, group_id: str) -> Group:
        """
        Get a group by ID.

        Args:
            group_id: The group ID to retrieve.

        Returns:
            The group.

        Raises:
            ValueError: If group_id is invalid.
            NotFoundError: If group not found.
            APIError: If API request fails.
        """
        self._validate_required(group_id, "group_id")

        try:
            # The API doesn't have a direct get-by-id endpoint,
            # so we use the list endpoint with id filter
            resp = self._groups_api.user_service_get_groups(
                limit=str(1),
                offset=str(0),
                ids=[group_id],
                external_group_ids=None,
                query=None,
            )

            result = getattr(resp, "result", None)
            if result is None or len(result) == 0:
                from taurus_protect.errors import NotFoundError

                raise NotFoundError(f"Group {group_id} not found")

            group = group_from_dto(result[0])
            if group is None:
                from taurus_protect.errors import NotFoundError

                raise NotFoundError(f"Group {group_id} not found")

            return _with_computed_flags(group)
        except Exception as e:
            from taurus_protect.errors import APIError, NotFoundError

            if isinstance(e, (APIError, NotFoundError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def list(
        self,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> Tuple[List[Group], Pagination]:
        """
        List groups, one page at a time.

        Args:
            limit: Page size (default 20, max 100).
            offset: Number of groups to skip; pass ``pagination.next_offset`` to continue.

        Returns:
            Tuple of (groups, pagination). A synthetic technical group can be appended
            beyond ``limit``; ``next_offset`` accounts for it.

        Raises:
            ValueError: If limit or offset are invalid.
            APIError: If API request fails.
        """
        page_size = resolve_page_size(limit, "limit")
        start = resolve_offset(offset)

        try:
            resp = self._groups_api.user_service_get_groups(**offset_query(page_size, start))

            rows = resp.result or []
            groups = [_with_computed_flags(g) for g in groups_from_dto(rows)]
            pagination = offset_pagination(
                PLUS_MIN_ROWS_LIMIT,
                limit=page_size,
                offset=start,
                served_rows=len(rows),
                total_items=resp.total_items,
                excluded=len(rows) - len(groups),
            )
            return groups, pagination
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e
